### ISSUE TITLE
Use AEAD's GCM / OCB instead of encryption+MAC.

### ISSUE THREAD

**Issue by vincenthz** [yesodweb/clientsession#23]
The AES-CTR+Skein-MAC method used by clientsession could be replaced by AES-GCM or AES-OCB, and in theory it would be much faster (practically current version have _lot_ of room for optimisations). Provided I send a PR for this, does clientsession need to remain backward compatible, and is that a good idea to provide different choosable alternatives to the user ?

**Comment by meteficha** [comment]
I just took a stab at using AES-GCM.  Unfortunately, the results weren't promising benchmark-wise.  You can see them here: https://gist.github.com/meteficha/a2bea768ef3255f8ab70.  At best AES-GCM was 28% slower, but on a unrealistic session size (80kb).  At worst it was 7.6x slower (0-byte session).

My CPU doesn't have AES-NI instructions, but the old version was using AES-CTR.  IIUC, AES-GCM should need just one more AES block encryption wrt AES-CTR.

My CPU also doesn't have the PCLMULQDQ instruction, perhaps this would make a huge difference?  Do most server CPUs have these instructions?

@vincenthz Do you think there's room for improvement on my stab?  Besides lowering to AES128, as it would make for a unfair benchmark :P.

**Comment by meteficha** [comment]
I forgot to mention, this is the commit: https://github.com/meteficha/clientsession/commit/e5960ee7e383dc7168f14df06335700a459950a9.   The benchmark is built with `--enable-tests`.  You can build the Skein one by taking the parent of that commit.

**Comment by vincenthz** [comment]
@meteficha: this is very likely the problem is on my side, as currently GCM is still quite slow compared to state-of-the-art implementations out there. My implementation of GCM doesn't use PCLMULQDQ as of yet; So it shouldn't make a difference overall in your benchmark that you don't have a cpu that support it.

GCM need one GFmul per blocks in addition to CTR mode, which is the place that is severely not optimised.

**Comment by meteficha** [comment]
@vincenthz I took a cursory look at your code and the only room for improvement I saw besides GFmul was removing SecureMem copies.  But at best this would bring improvements for the short plaintext case, the large plaintext case probably wouldn't be affected since there are a constant number of SecureMem copies.  IOW, AES-GCM would probably still be slower.  However, I'm not confident enough to try to optimize GFmul.

My current opinion is that using something more standard such as AES-GCM would be nice, but performance is the priority.  (After security, of course, but I AFAIK: Skein-MAC is secure, its reference implementation is bug-free, my bindings don't introduce any problems.)

**Comment by jckeme** [comment]
@vincenthz Hi, Sorry for bringing up this thread again. So did you end up implementing AES-OCB in any form? If yes, I am wondering how possible it might be to do this since there seems to be some liberty to use the AES-OCB for academic purposes.

**Comment by vincenthz** [comment]
@jckeme: yes AES-OCB has been implemented in cryptonite; also related ChaChaPoly1305 is available, which is my favorite default choice lately for this purpose.
